File: app/src/main/java/uk/co/transferx/app/ui/signup/fragment/SignUpStepThreeFragment.java
package uk.co.transferx.app.ui.signup.fragment;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.ColorRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.chaos.view.PinView;

import javax.inject.Inject;

import uk.co.transferx.app.ui.base.BaseFragment;
import uk.co.transferx.app.R;
import uk.co.transferx.app.TransferXApplication;
import uk.co.transferx.app.ui.mainscreen.MainActivity;
import uk.co.transferx.app.ui.signup.presenters.SignUpStepThreePresenter;
import uk.co.transferx.app.ui.tutorial.TutorialActivity;
import uk.co.transferx.app.ui.welcom.WelcomeActivity;

import static uk.co.transferx.app.util.Constants.EMAIL;
import static uk.co.transferx.app.util.Constants.EMPTY;
import static uk.co.transferx.app.util.Constants.PASSWORD;
import static uk.co.transferx.app.util.Constants.PIN_SHOULD_BE_INPUT;
import static uk.co.transferx.app.util.Constants.U_NAME;

/**
 * Created by smilevkiy on 15.11.17.
 */

public class SignUpStepThreeFragment extends BaseFragment implements SignUpStepThreePresenter.SignUpStepThreeUI {

    @Inject
    SignUpStepThreePresenter presenter;
    private PinView firstPinView, secondPinView;
    private TextView firsLabelPin, secondLabelPin;
    private ImageView backButton;
    private TextWatcher firstPinWatcher, secondPinWatcher;
    private boolean reEnterPin;

    @Override
    public String tagName() {
        return SignUpStepThreeFragment.class.getSimpleName();
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        ((TransferXApplication) getActivity().getApplication()).getAppComponent().inject(this);
        super.onCreate(savedInstanceState);
        Bundle bundle = getArguments();
        if (bundle != null) {
            presenter.setCredential(bundle.getString(U_NAME), bundle.getString(EMAIL), bundle.getString(PASSWORD));
            reEnterPin = bundle.getBoolean(PIN_SHOULD_BE_INPUT, false);
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.sign_up_step_three_fragment_layout, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        firstPinView = view.findViewById(R.id.first_pin);
        secondPinView = view.findViewById(R.id.second_pin);
        firsLabelPin = view.findViewById(R.id.first_label_pin);
        secondLabelPin = view.findViewById(R.id.second_label_pin);
        backButton = view.findViewById(R.id.back_button);
        backButton.setOnClickListener(v -> {
            hideKeyboard(backButton);
            getActivity().onBackPressed();
        });
        buttonNext = view.findViewById(R.id.sign_in);
        buttonNext.setText(reEnterPin ? R.string.sign_in : R.string.register);
        buttonNext.setOnClickListener(v -> presenter.validatePin());
        firstPinView.setAnimationEnable(true);
        secondPinView.setAnimationEnable(true);
    }

    @Override
    public void showErrorPin() {
        setStatusOfError(firstPinView, firsLabelPin, R.color.red);
        setStatusOfError(secondPinView, secondLabelPin, R.color.red);
        setLinesColor(R.color.red);
    }

    private void resetErrorPin() {
        setStatusOfError(firstPinView, firsLabelPin, R.color.black);
        setStatusOfError(secondPinView, secondLabelPin, R.color.black);
        setLinesColor(R.color.black);
    }

    @Override
    public void goToMainScreen() {
        Activity activity = getActivity();
        if (activity != null) {
            if (reEnterPin) {
                MainActivity.startMainActivity(activity);
                activity.finish();
                return;
            }
//            TutorialActivity.startTutorialActivity(activity);
            activity.finish();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.attachUI(this);
        firstPinView.addTextChangedListener(firstPinWatcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                resetErrorPin();
                presenter.setFirstPin(s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
        secondPinView.addTextChangedListener(secondPinWatcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                resetErrorPin();
                presenter.setSecondPin(s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }

    private void setLinesColor(@ColorRes int color) {
        firstPinView.setLineColor(ContextCompat.getColor(firsLabelPin.getContext(), color));
        secondPinView.setLineColor(ContextCompat.getColor(firsLabelPin.getContext(), color));
    }

    @Override
    public void onPause() {
        presenter.detachUI();
        firstPinView.removeTextChangedListener(firstPinWatcher);
        secondPinView.removeTextChangedListener(secondPinWatcher);
        firstPinView.setText(EMPTY);
        secondPinView.setText(EMPTY);
        super.onPause();
    }

    @Override
    public void showErrorFromBackend() {

    }

    @Override
    public void goToWelcome() {
        WelcomeActivity.startWelcomeActivity(getActivity());
    }

    @Override
    public void setButtonEnabled(boolean isEnabled) {
        setButtonStatus(isEnabled);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
    }
}

File: app/src/main/java/uk/co/transferx/app/ui/signup/fragment/SignUpStepThreeFragment.kt
package uk.co.transferx.app.ui.signup.fragment

import android.graphics.Color
import android.os.Bundle
import android.support.annotation.ColorRes
import android.support.design.widget.Snackbar
import android.support.v4.content.ContextCompat
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.sign_up_step_three_fragment_layout.*
import org.jetbrains.anko.startActivity
import uk.co.transferx.app.R
import uk.co.transferx.app.TransferXApplication
import uk.co.transferx.app.ui.base.BaseFragment
import uk.co.transferx.app.ui.landing.LandingActivity
import uk.co.transferx.app.ui.mainscreen.MainActivity
import uk.co.transferx.app.ui.signup.SignUpActivity
import uk.co.transferx.app.ui.signup.presenters.SignUpStepThreePresenter
import uk.co.transferx.app.util.Constants.*
import javax.inject.Inject

/**
 * Created by smilevkiy on 15.11.17.
 */

class SignUpStepThreeFragment : BaseFragment(), SignUpStepThreePresenter.SignUpStepThreeUI {
    @Inject
    lateinit var presenter: SignUpStepThreePresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        (activity?.application as TransferXApplication).appComponent.inject(this)
        super.onCreate(savedInstanceState)

        // Get credentials from previous steps
        presenter.setCredential(arguments!!.getString(U_NAME), arguments!!.getString(EMAIL),
                arguments!!.getString(PASSWORD), arguments!!.getString(PHONE_NUMBER), arguments!!.getString(COUNTRY))
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.sign_up_step_three_fragment_layout, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        buttonBack.setOnClickListener {
            hideKeyboard(buttonBack)
            activity?.onBackPressed()
        }

        buttonNext = buttonSignUp
        buttonNext.setOnClickListener {
            hideKeyboard(it)
            presenter.signUpUser()
        }

        firstPinEntry.setAnimationEnable(true)
        secondPinEntry.setAnimationEnable(true)
    }

    override fun onResume() {
        super.onResume()
        presenter.attachUI(this)
        firstPinEntry.addTextChangedListener(firstPinWatcher)
        secondPinEntry.addTextChangedListener(secondPinWatcher)
    }

    override fun onPause() {
        presenter.detachUI()
        firstPinEntry.removeTextChangedListener(firstPinWatcher)
        secondPinEntry.removeTextChangedListener(secondPinWatcher)
        firstPinEntry.setText(EMPTY)
        secondPinEntry.setText(EMPTY)
        super.onPause()
    }

    override fun showErrorPin() {
        firstPinEntry.setText(EMPTY)
        secondPinEntry.setText(EMPTY)
        setLinesColor(R.color.red)
        firstPinEntry.requestFocus()

        val snackbar = Snackbar.make(view!!, getString(R.string.no_match_pin), Snackbar.LENGTH_LONG)
        snackbar.view.setBackgroundColor(Color.RED)
        snackbar.show()
    }

    private fun resetErrorPin() {
        setLinesColor(R.color.black)
    }

    override fun goToMainScreen() {
        MainActivity.startMainActivity(activity)
        activity?.finish()
    }

    override fun goToConfirmationScreen() {
        (activity as SignUpActivity).showNextOrPreviousFragment(3, null)
    }

    override fun showErrorFromBackend() {
    }

    override fun setButtonStatus(isEnabled: Boolean) {
        buttonNext.isEnabled = isEnabled
        buttonNext.background = if (isEnabled) ContextCompat.getDrawable(context!!, R.drawable.btn_style_bright)
        else ContextCompat.getDrawable(context!!, R.drawable.btn_disabled)
    }

    override fun goToWelcome() {
        context?.startActivity<LandingActivity>()
    }

    override fun setButtonEnabled(isEnabled: Boolean) {
        setButtonStatus(isEnabled)
    }

    override fun tagName(): String {
        return SignUpStepThreeFragment::class.java.simpleName
    }

    private fun setLinesColor(@ColorRes color: Int) {
        firstPinEntry.setLineColor(ContextCompat.getColor(context!!, color))
        secondPinEntry.setLineColor(ContextCompat.getColor(context!!, color))
    }

    private val firstPinWatcher = object : TextWatcher {
        override fun afterTextChanged(s: Editable?) {
        }

        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        }

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            s?.let { presenter.setFirstPin(s.toString()) }
            resetErrorPin()
        }

    }

    private val secondPinWatcher = object : TextWatcher {
        override fun afterTextChanged(s: Editable?) {
        }

        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        }

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            s?.let { presenter.setSecondPin(s.toString()) }
            resetErrorPin()
        }

    }

}
